cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,298 @@
+	}
+	if *clusterInfo.Spec.DatastoreReady {
+		*clusterInfo.Spec.DatastoreReady = false
+		if clusterInfo, err = c.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
+			return fmt.Errorf("failed to disable cluster: %s", err)
+		}
+	}
+
+	// Also disable cni by disabling the binaries.
+	for _, binary := range binariesToDisable {
+		dir, file := path.Split(binary)
+		disabledBinary := path.Join(dir, file+".disabled")
+		if err = os.Rename(binary, disabledBinary); err != nil {
+			return fmt.Errorf("failed to disable binary %s: %s", binary, err)
+		}
+	}
+
+	podList, err := k8sClient.CoreV1().Pods("").List(metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodename),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list pods: %s", err)
+	}
+	podIPMap := make(map[string]*corev1.Pod)
+	for i := 0; i < len(podList.Items); i++ {
+		podIPMap[podList.Items[i].Status.PodIP] = &podList.Items[i]
+	}
+
+	// Read in all the files in the host-local directory.
+	files, err := ioutil.ReadDir(ipAllocPath)
+	if err != nil {
+		return fmt.Errorf("failed to read path %s: %s", ipAllocPath, err)
+	}
+
+	// For each file, convert it into an IP allocation and then delete the file.
+	for _, f := range files {
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") || f.Name() == "lock" {
+			if err = os.Remove(fname); err != nil {
+				return fmt.Errorf("failed to remove file %s: %s", fname, err)
+			}
+			continue
+		}
+
+		// The name of the file is its IP address.
+		ip, _, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name()))
+		if err != nil {
+			return fmt.Errorf("failed to parse IP %s: %s", f.Name(), err)
+		}
+
+		// The contents are the container ID.
+		b, err := ioutil.ReadFile(fname)
+		if err != nil {
+			return fmt.Errorf("failed to read file %s: %s", fname, err)
+		}
+		containerID := string(b)
+
+		// See if the IP hasn't already been assigned for the handleID.
+		handleID := utils.GetHandleID("k8s-pod-network", containerID)
+		/*
+			ips, err := c.IPAM().IPsByHandle(ctxt, handleID)
+			if err == nil {
+			} else if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+				// Return error if it is NOT does not exist, otherwise continue.
+				return fmt.Errorf("failed to fetch ips for handleID %s: %s", handleID, err)
+			}
+		*/
+
+		// Get the pod resource associated with the IP.
+		pod, ok := podIPMap[f.Name()]
+		if !ok {
+			log.WithField("ip", f.Name()).Info("pod not found for IP, deleting and continuing...")
+			if err = os.Remove(fname); err != nil {
+				return fmt.Errorf("failed to remove file %s: %s", fname, err)
+			}
+			continue
+		}
+
+		// Store allocation to datastore
+		if err = c.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+			IP:       *ip,
+			HandleID: &handleID,
+			Hostname: nodename,
+			Attrs: map[string]string{
+				"node":      nodename,
```

Comment:
Just added some constants for these which would be good to use: https://github.com/projectcalico/libcalico-go/pull/1035/files